code/controller/astar.cpp

```diff
@@ -0,0 +1,109 @@
+#include "astar.h"
+#include <cmath>
+#include <set>
+#include <queue>
+#include <iostream>
+
+using namespace std;
+
+Astar::Astar() {}
+
+double Astar::manhattanDist(Astar::Coord cur, Astar::Coord dest) {
+    return abs(dest.x - cur.x) + abs(dest.y - cur.y);
+}
+
+bool Astar::isValid(int x, int y, int maxRow, int maxCol) {
+    return (x >= 0) && (x < maxRow) && (y >= 0) && (y < maxCol);
+}
+
+vector<Astar::Coord> Astar::getNeighbours(Astar::Coord c, int maxRow, int maxCol) {
+    vector<Astar::Coord> neighbours;
+    int cur_x = c.x;
+    int cur_y = c.y;
+    if (isValid(cur_x-1, cur_y, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x-1, cur_y});
+    }
+    if (isValid(cur_x+1, cur_y, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x+1, cur_y});
+    }
+    if (isValid(cur_x, cur_y-1, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x, cur_y-1});
+    }
+    if (isValid(cur_x, cur_y+1, maxRow, maxCol)) {
+        neighbours.push_back(Astar::Coord{cur_x, cur_y+1});
+    }
+
+    return neighbours;
+}
+
+// For testing purposes
+Astar::grid Astar::makeTerrain() {
```

Comment:
add this to `test/controller/astar_test.cpp` and see other files for testing examples